Bug: race condition in squashfuse mount path - command not found

### stacker version

v1.0.0-rc1

### Describe the bug

I see this problem occasionally when building bootkit and have seen it elsewhere as well.
[here](https://github.com/project-machine/bootkit/actions/runs/4316377353/jobs/7532358700) is a github-actions reproduce.

The problem manifests itself like:
```
2023-03-02T17:01:18.1854206Z preparing image stubby-build-env...
2023-03-02T17:01:18.1987179Z couldn't find AppArmor profile lxc-container-default-cgns
2023-03-02T17:01:18.2097611Z + pkgtool install binutils gcc gnu-efi libc6-dev make tar
2023-03-02T17:01:18.2098345Z /stacker/.stacker-run.sh: 2: pkgtool: not found
2023-03-02T17:01:18.2153865Z error: run commands failed: execute failed: exit status 127
2023-03-02T17:01:18.2233898Z error: exit status 1
2023-03-02T17:01:18.2238643Z make: *** [Makefile:8: build] Error 1
```

the 'not found' file lives in a layer 'minbase'. they look like this:

 - docker://ubuntu:jammy
 - minbase (does an import pkgtool and runs 'pkgtool upgrade' and 'pkgtool install ca-certificates'
 - other layer (stubby-build-env) based on minbase, runs 'pkgtool' as the first thing in its build process.

bootkit builds with 'stacker recursive-build'.
here is an attempt at a stacker.yaml that would reproduce, but I have not tried it. pkgtool could really be any script.


```yaml
minbase:
  build_only: true
  from:
    type: docker
    url: ${{DOCKER_BASE}}ubuntu:jammy
  import:
    - pkgtool
  run: |
    mirror=${{UBUNTU_MIRROR}}

    rel=$(. /etc/os-release && echo "${VERSION_CODENAME}")
    [ -n "$rel" ]

    cp /etc/apt/sources.list /etc/apt/sources.list.dist
    cat > /etc/apt/sources.list <<EOF
    deb [by-hash=force] $mirror ${rel} main universe
    deb [by-hash=force] $mirror ${rel}-updates main universe
    deb [by-hash=force] $mirror ${rel}-security main universe
    EOF

    cp /stacker/pkgtool /usr/local/bin

    #pkgtool upgrade
    #pkgtool install ca-certificates

rootfs:
  from:
    type: built
    tag: minbase
  run: |
    pkgtool install ca-certificates
```

### To reproduce

I see the issue occasionally just using 'make' in https://github.com/project-machine/bootkit.
I have only seen it once or twice on github actions before now.

### Expected behavior

_No response_

### Screenshots

_No response_

### Additional context

_No response_

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -33,6 +33,7 @@ jobs:
           go-version: ${{ matrix.go-version }}
       - name: install dependencies
         run: |
+          sudo add-apt-repository -y ppa:project-machine/squashfuse
           sudo apt-get update
           sudo apt-get install -yy lxc-utils lxc-dev libacl1-dev jq libcap-dev libseccomp-dev libpam-dev bats parallel libzstd-dev
           GO111MODULE=off go get github.com/opencontainers/umoci/cmd/umoci
@@ -41,6 +42,7 @@ jobs:
           sudo apt-get install -yy autoconf automake make autogen autoconf libtool binutils git squashfs-tools libcryptsetup-dev libdevmapper-dev cryptsetup-bin squashfuse
           (cd /tmp && git clone https://github.com/AgentD/squashfs-tools-ng && cd squashfs-tools-ng && ./autogen.sh && ./configure --prefix=/usr && make -j2 && sudo make -j2 install && sudo ldconfig -v)
           (cd /tmp && git clone https://github.com/anuvu/squashfs && cd squashfs && make && sudo cp squashtool/squashtool /usr/bin)
+          echo "running kernel is: $(uname -a)"
       - if: github.event_name != 'release' || github.event.action != 'published'
         name: Build
         run: |

--- a/pkg/squashfs/squashfs.go
+++ b/pkg/squashfs/squashfs.go
@@ -13,6 +13,7 @@ import (
 	"strings"
 	"sync"
 	"syscall"
+	"time"
 
 	"github.com/pkg/errors"
 	"golang.org/x/sys/unix"
@@ -211,7 +212,7 @@ func maybeKernelSquashMount(squashFile, extractDir string) (bool, error) {
 	}
 
 	// we can't really tell why the mount failed. mount(8) does not give a lot specific rc exits.
-	log.Debugf("maybeKernelSquashMount(%s) exited %d: %s\n", squashFile, status.ExitStatus(), output.String())
+	log.Debugf("maybeKernelSquashMount(%s) exited %d: %s", squashFile, status.ExitStatus(), strings.TrimRight(output.String(), "\n"))
 	return false, kernelSquashMountFailed
 }
 
@@ -224,10 +225,14 @@ func findSquashfusePath() string {
 
 var squashNotFound = errors.Errorf("squashfuse program not found")
 
+// squashFuse - mount squashFile to extractDir
+// return a pointer to the squashfuse cmd.
+// The caller of the this is responsible for the process created.
 func squashFuse(squashFile, extractDir string) (*exec.Cmd, error) {
 	sqfuse := findSquashfusePath()
+	var cmd *exec.Cmd
 	if sqfuse == "" {
-		return nil, squashNotFound
+		return cmd, squashNotFound
 	}
 
 	// given extractDir of path/to/some/dir[/], log to path/to/some/.dir-squashfs.log
@@ -235,26 +240,61 @@ func squashFuse(squashFile, extractDir string) (*exec.Cmd, error) {
 
 	var cmdOut io.Writer
 	var err error
-	var nilCmd *exec.Cmd
 
 	logf := filepath.Join(path.Dir(extractDir), "."+filepath.Base(extractDir)+"-squashfuse.log")
 	if cmdOut, err = os.OpenFile(logf, os.O_RDWR|os.O_TRUNC|os.O_CREATE, 0644); err != nil {
 		log.Infof("Failed to open %s for write: %v", logf, err)
-		return nilCmd, err
+		return cmd, err
+	}
+
+	fiPre, err := os.Lstat(extractDir)
+	if err != nil {
+		return cmd, errors.Wrapf(err, "Failed stat'ing %q", extractDir)
+	}
+	if fiPre.Mode()&os.ModeSymlink != 0 {
+		return cmd, errors.Errorf("Refusing to mount onto a symbolic linkd")
 	}
 
 	// It would be nice to only enable debug (or maybe to only log to file at all)
 	// if 'stacker --debug', but we do not have access to that info here.
 	// to debug squashfuse, use "allow_other,debug"
-	cmd := exec.Command(sqfuse, "-f", "-o", "allow_other,debug", squashFile, extractDir)
+	cmd = exec.Command(sqfuse, "-f", "-o", "allow_other,debug", squashFile, extractDir)
 	cmd.Stdin = nil
 	cmd.Stdout = cmdOut
 	cmd.Stderr = cmdOut
 	cmdOut.Write([]byte(fmt.Sprintf("# %s\n", strings.Join(cmd.Args, " "))))
-	log.Debugf("Extracting %s -> %s with squashfuse [%s]", squashFile, extractDir, logf)
+	log.Debugf("Extracting %s -> %s with %s [%s]", squashFile, extractDir, sqfuse, logf)
 	err = cmd.Start()
 	if err != nil {
-		return nilCmd, err
+		return cmd, err
+	}
+
+	// now poll/wait for one of 3 things to happen
+	// a. child process exits - if it did, then some error has occurred.
+	// b. the directory Entry is different than it was before the call
+	//    to sqfuse.  We have to do this because we do not have another
+	//    way to know when the mount has been populated.
+	//    https://github.com/vasi/squashfuse/issues/49
+	// c. a timeout (timeLimit) was hit
+	startTime := time.Now()
+	timeLimit := 30 * time.Second
+	go func() {
+		cmd.Wait()
+	}()
+	for count := 0; !fileChanged(fiPre, extractDir); count++ {
+		if cmd.ProcessState != nil {
+			// process exited, the Wait() call in the goroutine above
+			// caused ProcessState to be populated.
+			return cmd, errors.Errorf("squashFuse mount of %s with %s exited unexpectedly with %d", squashFile, sqfuse, cmd.ProcessState.ExitCode())
+		}
+		if time.Since(startTime) > timeLimit {
+			cmd.Process.Kill()
+			return cmd, errors.Wrapf(err, "Gave up on squashFuse mount of %s with %s after %s", squashFile, sqfuse, timeLimit)
+		}
+		if count%10 == 1 {
+			log.Debugf("%s is not yet mounted...(%s)", extractDir, time.Since(startTime))
+		}
+		time.Sleep(time.Duration(50 * time.Millisecond))
 	}
 
 	return cmd, nil
@@ -272,8 +312,13 @@ func ExtractSingleSquash(squashFile string, extractDir string, storageType strin
 		return err
 	}
 
-	_, err = squashFuse(squashFile, extractDir)
-	if err == nil || err != squashNotFound {
+	cmd, err := squashFuse(squashFile, extractDir)
+	if err == nil {
+		if err := cmd.Process.Release(); err != nil {
+			return errors.Errorf("Failed to release process %s: %v", cmd, err)
+		}
+		return nil
+	} else if err != squashNotFound {
 		return err
 	}
 	if p := which("unsquashfs"); p != "" {

--- a/pkg/squashfs/verity.go
+++ b/pkg/squashfs/verity.go
@@ -71,7 +71,6 @@ import (
 	"strconv"
 	"strings"
 	"syscall"
-	"time"
 	"unsafe"
 
 	"github.com/anuvu/squashfs"
@@ -79,7 +78,6 @@ import (
 	"github.com/martinjungblut/go-cryptsetup"
 	"github.com/pkg/errors"
 	"golang.org/x/sys/unix"
-	"stackerbuild.io/stacker/pkg/log"
 	"stackerbuild.io/stacker/pkg/mount"
 )
 
@@ -229,30 +227,13 @@ func GuestMount(squashFile string, mountpoint string) error {
 	}
 	mountpoint = abs
 
-	fiPre, err := os.Lstat(mountpoint)
-	if err != nil {
-		return errors.Wrapf(err, "Failed stat'ing %q", mountpoint)
-	}
-	if fiPre.Mode()&os.ModeSymlink != 0 {
-		return errors.Errorf("Refusing to mount onto a symbolic linkd")
-	}
-
 	cmd, err := squashFuse(squashFile, mountpoint)
 	if err != nil {
 		return err
 	}
-	err = cmd.Process.Release()
-	if err != nil {
-		return errors.Wrapf(err, "Failed releasing squashfuse process")
+	if err := cmd.Process.Release(); err != nil {
+		return errors.Errorf("Failed to release process after guestmount %s: %v", squashFile, err)
 	}
-
-	for count := 0; !fileChanged(fiPre, mountpoint); count++ {
-		if count%10 == 0 {
-			log.Debugf("%s is not yet mounted...\n", mountpoint)
-		}
-		time.Sleep(time.Duration(100 * time.Millisecond))
-	}
-
 	return nil
 }
 

